loadWithXhr response handling

klim705 opened the pull request:
Add a workaround for `xhr.responseType = 'json'` for IE, which doesn't support that type. Additionally, add checks before attempting to access `xhr.responseXML` or `xhr.responseText`, which throws if `responseType` isn't an appropriate type.

`loadWithXhrSpec` currently only has tests for data URI loading. Is there a good way to test sending/loading of data without passing an actual URL? Mocking/spying doesn't seem possible as is since the request isn't publicly accessible.

mramato commented on Source/Core/loadWithXhr.js:
```diff
@@ -152,12 +152,20 @@ define([
                 if (defined(xhr.response)) {
-                    deferred.resolve(xhr.response);
+                    if (responseType === 'json' && typeof xhr.response === 'string') { //IE doesn't support responseType = 'json'
+                        try{
```
Spacing.  Can you run the autoformatter on this new block of code please.

mramato commented on Source/Core/loadWithXhr.js:
```diff
@@ -152,12 +152,20 @@ define([
+                        try{
+                            deferred.resolve(JSON.parse(xhr.response));
+                        } catch (e) {
+                            deferred.reject(new RuntimeError('unknown XMLHttpRequest response type.'));
```
I think this should just be `deferred.reject(e)` so that whatever the real browser error is comes though.

mramato commented:
This branch is ancient compared to master (even though it would merge cleanly).  Can you please merge in AGI master so that it's up to date.

mramato commented:
> loadWithXhrSpec currently only has tests for data URI loading. Is there a good way to test sending/loading of data without passing an actual URL? Mocking/spying doesn't seem possible as is since the request isn't publicly accessible.

It's okay to actually have the test request data, just use a relative path.  If you just need a non-JSON file, you can request anything from `Specs/Data`.  If there's a specific file you need, just put it into `Specs/Data`.

mramato commented:
Thanks @klim705!

Specifically, we ran into the problem where the server ended up returning HTML when JSON was expected and `loadWithXhr` would never reject the promise.  

@kring or @shunter, can you do a quick review.  You guys probably know the edge cases here a little better than I do.

klim705 commented:
Merged AGI's master, ran the autoformatter, and switched the catch block's `deferred` to reject with the exception. Also added unit tests for URL loading following a format similar to in `Core/loadTextSpec` to mock loading.

pjcozzi commented:
Does this warrant an update to CHANGES.md?

mramato commented on Specs/Core/loadWithXhrSpec.js:
```diff
@@ -133,18 +143,364 @@ defineSuite([
+        var fakeXHR;
+
+        beforeEach(function() {
+            fakeXHR = jasmine.createSpyObj('XMLHttpRequest', ['send', 'open', 'setRequestHeader', 'abort', 'getAllResponseHeaders']);
```
Creating a completely fake spy object here (at least partially) invalidates the entire test.  The crux of the problem is that XMLHttpRequest has different behaviors in different browsers, so mocking it to be consistent in all cases isn't actually testing what we need to test.

mramato commented on Source/Core/loadWithXhr.js:
```diff
@@ -153,12 +153,20 @@ define([
             if (xhr.status >= 200 && xhr.status < 300) {
                 if (defined(xhr.response)) {
-                    deferred.resolve(xhr.response);
+                    if (responseType === 'json' && typeof xhr.response === 'string') { //IE doesn't support responseType = 'json'
```
I spent a bunch of time verifying browser behaviors here and ended up refacotring `xhr.onload` to incorporate your changes but also better separate "old browser" behavior from modern browsers.  Can you replace xhr.onload with the below?

```javascript
        xhr.onload = function() {
            if (xhr.status < 200 || xhr.status >= 300) {
                deferred.reject(new RequestErrorEvent(xhr.status, xhr.response, xhr.getAllResponseHeaders()));
                return;
            }

            var response = xhr.response;
            var browserResponseType = xhr.responseType;

            //All modern browsers will go into either the first if block or last else block.
            //Other code paths support older browsers that either do not support the supplied responseType
            //or do not support the xhr.response property.
            if (defined(response) && (!defined(responseType) || (browserResponseType === responseType))) {
                deferred.resolve(response);
            } else if ((responseType === 'json') && typeof response === 'string') {
                try {
                    deferred.resolve(JSON.parse(response));
                } catch (e) {
                    deferred.reject(e);
                }
            } else if ((browserResponseType === '' || browserResponseType === 'document') && defined(xhr.responseXML) && xhr.responseXML.hasChildNodes()) {
                deferred.resolve(xhr.responseXML);
            } else if ((browserResponseType === '' || browserResponseType === 'text') && defined(xhr.responseText)) {
                deferred.resolve(xhr.responseText);
            } else {
                deferred.reject(new RuntimeError('Invalid XMLHttpRequest response type.'));
            }
        };
```

This will also break a couple of your tests, but that's because your tests are depending on non-standard XHR behavior that is a result of your mocking the XHR object.  I spoke with @shunter offline and he disagreed with me and felt that mocking XHR was okay, but we need to make sure the behavior matches what modern browsers do (and also matches old browser behavior in tests that exercise it.).

So update your tests to work with the above code and then I'll take another look.

klim705 replied:
To verify - mocking XHR is okay for testing or no?

mramato replied:
It's okay, but I would only do it for things that can't easily be done with actual XHR requests and make sure the behavior matches what browsers do.  For example, in all browsers, `responseType` defaults to an empty string and setting an unsupported type just sets `responseType` back to an empty string.

mramato commented:
> Does this warrant an update to CHANGES.md?

This is a trivial edge case that we just happened to never encounter, so probably not.  I doubt any of our users call `loadWithXhr` directly and higher level functions (like loadJson) did not have this problem.

klim705 commented:
Ready for another look.

mramato commented on Specs/Core/loadWithXhrSpec.js:
```diff
@@ -133,18 +143,269 @@ defineSuite([
+            it('results in an HTTP status code greater than or equal to 300', function() {
+                return loadWithXhr({
+                    url : 'http://example.invalid'
+                }).then(function(resolvedValue) {
```
Don't use `.then(resolveCallback, rejectCallback)`  Instead use `.then(resolveCallback).otherwise(rejectCallback).  The former does not catch all exceptions, while the latter will.

mramato commented on Specs/Core/loadWithXhrSpec.js:
```diff
@@ -133,18 +143,269 @@ defineSuite([
+                return loadWithXhr({
+                    url : 'http://example.invalid'
+                }).then(function(resolvedValue) {
+                    expect(resolvedValue).toBeUndefined();
```
This should code should never be reached, right?  In that case this should be `fail('expected promise to reject');`

mramato commented on Specs/Core/loadWithXhrSpec.js:
```diff
@@ -133,18 +143,269 @@ defineSuite([
+                }).then(function(resolvedValue) {
+                    expect(resolvedValue).toBe('CesiumBoxTest-NoTechnique.gltf is a modified glTF that has techniques, shaders & programs removed.');
+                }, function(rejectedError) {
+                    expect(rejectedError).toBeUndefined();
```
There's no need to have a reject handler if you don't expect the promise to be rejected.  Cesium's scaffolding will automatically handle rejected promises and report an error.

klim705 commented:
Ready.

mramato commented:
Thanks @klim705
